serde-generate/tests/lcs_runtime.rs

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) Facebook, Inc. and its affiliates
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use libra_canonical_serialization as lcs;
+use serde::{Deserialize, Serialize};
+use serde_generate::{python3, rust, test_utils};
+use serde_reflection::{Registry, Result, Samples, Tracer, TracerConfig};
+use std::fs::File;
+use std::io::Write;
+use std::process::Command;
+use tempfile::tempdir;
+
+#[derive(Serialize, Deserialize)]
+struct Test {
+    a: Vec<u64>,
+    b: (u32, u32),
+    c: Choice,
+}
+
+#[derive(Serialize, Deserialize)]
+enum Choice {
+    A,
+    B(u64),
+    C { x: u8 },
+}
+
+fn get_local_registry() -> Result<Registry> {
+    let mut tracer = Tracer::new(TracerConfig::default());
+    let samples = Samples::new();
+    tracer.trace_type::<Test>(&samples)?;
+    tracer.trace_type::<Choice>(&samples)?;
+    Ok(tracer.registry()?)
+}
+
+#[test]
+fn test_python_lcs_runtime_on_simple_data() {
+    let registry = get_local_registry().unwrap();
+    let dir = tempdir().unwrap();
+    let source_path = dir.path().join("test.py");
+    let mut source = File::create(&source_path).unwrap();
+    python3::output(&mut source, &registry).unwrap();
+
+    let reference = lcs::to_bytes(&Test {
+        a: vec![4, 6],
+        b: (3, 5),
+        c: Choice::C { x: 7 },
+    })
+    .unwrap();
+    writeln!(
+        source,
+        r#"
+import lcs
+
+value = Test([4, 6], (3, 5), Choice.C(7))
+
+s = lcs.serialize(value, Test)
+assert s == bytes.fromhex("{}")
+
+v, buffer = lcs.deserialize(s, Test)
+assert len(buffer) == 0
+assert v == value
+assert v.c.x == 7
+
+v.b = (3, 0)
+t = lcs.serialize(v, Test)
+assert len(t) == len(s)
+assert t != s
+"#,
+        hex::encode(&reference),
+    )
+    .unwrap();
+
+    let python_path = std::env::var("PYTHONPATH").unwrap_or_default() + ":runtime/python";
+    let output = Command::new("python3")
+        .arg(source_path)
+        .env("PYTHONPATH", python_path)
+        .output()
+        .unwrap();
+
+    std::io::stdout().write_all(&output.stdout).unwrap();
+    std::io::stderr().write_all(&output.stderr).unwrap();
+    assert_eq!(String::new(), String::from_utf8_lossy(&output.stderr));
+    assert!(output.status.success());
+}
+
+#[test]
+fn test_python_lcs_runtime_on_all_supported_types() {
+    let registry = test_utils::get_registry().unwrap();
+    let dir = tempdir().unwrap();
+    let source_path = dir.path().join("test.py");
+    let mut source = File::create(&source_path).unwrap();
+    python3::output(&mut source, &registry).unwrap();
+
+    let values = test_utils::get_sample_values();
```

Comment:
@thefallentree What you are suggesting is happening here and below. The values are not random at the moment but they cover every LCS construction.

Reply:
We test the installer separately (verifying that moving files around does not break python import).

Reply:
This being said, we could do better on testing. In particular, it's missing negative examples.